util/src/main/java/com/google/cloud/hadoop/util/CredentialFactory.java

```diff
@@ -287,6 +297,77 @@ public Credential getCredentialFromJsonKeyFile(
     }
   }
 
+  /**
+   *
+   * @param credentialProviderPath
+   * @return Correct Credential provider on basis of path configured
+   * @throws IOException
+   */
+  private CredentialProvider getHadoopCredentialProvider(String credentialProviderPath) throws IOException {
+    CredentialProvider result = null;
+    try {
+      URI uri = new URI(credentialProviderPath);
+      boolean found = false;
+      for(CredentialProviderFactory factory: serviceLoader) {
+        CredentialProvider kp = factory.createProvider(uri, new Configuration());
+        if (kp != null) {
+          result = kp;
+          found = true;
+        }
+      }
+      if (!found) {
+        throw new IOException("No CredentialProviderFactory for " + uri + " in " +
+                credentialProviderPath);
+      }
+    } catch (URISyntaxException error) {
+      throw new IOException("Bad configuration of " + credentialProviderPath, error);
+    }
+
+    return result;
+  }
+
+  /**
+   * Initializes OAuth2 credential from a Hadoop credential Provider path.
+   * More details around Hadoop credential provider
+   * <a href="https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/CredentialProviderAPI.html">
+   * Hadoop Credential </a>
+   *
+   * @param credentialProviderPath Credential provider.
+   * @param scopes List of well-formed desired scopes to use with the credential.
+   * @param transport The HttpTransport used for authorization
+   */
+  public Credential getCredentialFromCredentialProvider(
```

Comment:
Will it be easier just to use [Configuration#getPassword](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/conf/Configuration.html#getPassword(java.lang.String)) instead?

Reply:
Currently, Configuration#getPassword create/initializes credential provider for each call. In this case, if we have n entries, each call will try to load "n" number of times, hence I kept it like this.

Reply:
Yes, but it should be called only once per GHFS instantiation, not for every API call, isn't it?

Reply:
Yes, that's right. This will go only during GHFS instantiation. "n" meant to be the number of attributes in credentials (private_key, private_key_id, .... email etc), which will also happen in all the mappers. If we are fine with that, I can make the change to use Configuration#getPassword.

Reply:
@medb - can you let me know your thoughts?

Reply:
I'm inclined toward using Configuration#getPassword, but I plan to refactor credential-related code to be able to use Configuration#getPassword [here](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/cf5c004c4e29dd487cc5e9ec185c0b2d60a06fc6/util/src/main/java/com/google/cloud/hadoop/util/EntriesCredentialConfiguration.java#L303).

This refactoring could impact this PR, that's why I propose to wait for couple days for refactoring, before making decision on this PR.

Reply:
Sure. Thanks for letting me know on this.

Reply:
Sorry for the delay, started to work on it but it got deprioritized, will try to finish it in next couple days.